src/bundle/Security/Http/EventListener/CheckTwoFactorCodeReuseListener.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scheb\TwoFactorBundle\Security\Http\EventListener;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorCodeCheckEvent;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorCodeReusedEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use function assert;
+use function sha1;
+
+/**
+ * @final
+ */
+class CheckTwoFactorCodeReuseListener implements EventSubscriberInterface
+{
+    public const LISTENER_PRIORITY = 0;
+
+    public function __construct(
+        private readonly EventDispatcherInterface $eventDispatcher,
+        private readonly mixed $cache = null,
+    ) {
+    }
+
+    public function checkForCodeReuse(TwoFactorCodeCheckEvent $event): void
+    {
+        if (null === $this->cache) {
+            return;
+        }
+
+        $cacheItem = null;
+        assert($cacheItem instanceof CacheItemInterface || null === $cacheItem);
+        $cacheKey = 'scheb_two_factor_code_reuse.'
+            .sha1($event->getUser()->getUserIdentifier().'.'.$event->getCode());
```

Comment:
As the useridentifier can contain characters that are not allowed I decided to create a hash from the user-identifier and the code. The result is the same, a unique key for a combination of username and 2FA-code. 

The added benefit is, that neither value will be exposed in the cache so from a security point of view this is better than using the username and the token.